### PR TITLE
Allow savefig to save SVGs on FIPS enabled systems #18192

### LINKED ISSUE #18192: Cannot save SVG file with FIPS compliant Python
### Bug report

**Bug summary**

If you use a FIPS (Federal Information Processing Standards) compliant Python install, then you cannot use `plt.savefig` to save an image in SVG format. This is because the `RenderSVG._make_id` method takes the first 10 characters of a `hashlib.md5` digest of entries in the SVG as the ID for that part of the SVG file. The MD5 algorithm is disabled on FIPS compliant systems or in FIPS compliant Python installs.

**Code for reproduction**

The big requirement here is a FIPS compliant system or at least a Python install compiled to be FIPS compliant. Then the following code will reproduce the error.

To reproduce, 
1. I started a CentOS 8 virtual machine
2. Ran `sudo fips-mode-setup --enable && sudo reboot`
3. After the reboot I confirmed FIPS mode had been enabled by running `fips-mode-setup --check` and it returned `FIPS mode is enabled.`.
4. `sudo dnf install python3`
5. `python3 -m venv venv`
6. `./venv/bin/python -m pip install -U pip setuptools`
7. `./venv/bin/python -m pip install matplotlib`
8. `./venv/bin/python demo.py`. The contents of `demo.py` are listed below.

```python
""" demo.py: demonstrates that matplotlib can't save SVGs if FIPS mode enabled """
import matplotlib.pyplot as plt
plt.plot(range(10))
plt.savefig(range10.png")
plt.savefig("range10.svg")
```

**Actual outcome**

```
Traceback (most recent call last):
  File "demo.py", line 5, in <module>
    plt.savefig("range10.svg")
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/pyplot.py", line 842, in savefig
    res = fig.savefig(*args, **kwargs)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/figure.py", line 2311, in savefig
    self.canvas.print_figure(fname, **kwargs)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/backend_bases.py", line 2217, in print_figure
    **kwargs)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/backends/backend_svg.py", line 1318, in print_svg
    self._print_svg(filename, fh, **kwargs)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/backend_bases.py", line 1639, in wrapper
    return func(*args, **kwargs)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/backends/backend_svg.py", line 1342, in _print_svg
    self.figure.draw(renderer)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/artist.py", line 41, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/figure.py", line 1864, in draw
    renderer, self, artists, self.suppressComposite)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/image.py", line 132, in _draw_list_compositing_images
    a.draw(renderer)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/artist.py", line 41, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/cbook/deprecation.py", line 411, in wrapper
    return func(*inner_args, **inner_kwargs)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/axes/_base.py", line 2748, in draw
    mimage._draw_list_compositing_images(renderer, self, artists)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/image.py", line 132, in _draw_list_compositing_images
    a.draw(renderer)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/artist.py", line 41, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/axis.py", line 1169, in draw
    tick.draw(renderer)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/artist.py", line 41, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/axis.py", line 291, in draw
    artist.draw(renderer)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/artist.py", line 41, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/lines.py", line 854, in draw
    fc_rgba)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/backends/backend_svg.py", line 683, in draw_markers
    oid = self._make_id('m', dictkey)
  File "/home/daytonb/venv/lib64/python3.6/site-packages/matplotlib/backends/backend_svg.py", line 440, in _make_id
    m = hashlib.md5()
ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS
```

**Expected outcome**

Expect no errors and an SVG named "range10.svg" that shows the same plot as a PNG named "range10.png".

**Matplotlib version**

  * Operating system: CentOS Linux release 8.2.2004 (Core)
     * `fips-mode-setup --check` returns `FIPS mode is enabled`
  * Matplotlib version: 3.3.0
  * Matplotlib backend (`print(matplotlib.get_backend())`): agg
  * Python version: 3.6.8
  * Jupyter version (if applicable): NA
  * Other libraries: 
```
[daytonb@centos-s-1vcpu-1gb-nyc1-01 ~]$ venv/bin/python -m pip list
Package         Version
--------------- -------
cycler          0.10.0
kiwisolver      1.2.0
matplotlib      3.3.0
numpy           1.19.1
Pillow          7.2.0
pip             20.2.1
pyparsing       2.4.7
python-dateutil 2.8.1
setuptools      49.2.1
six             1.15.0
```

### PR DESCRIPTION
**Pull request by daytonb**
## PR Summary

This pull request resolves #18192.

If you use a FIPS (Federal Information Processing Standards) enabled system, then you cannot use `plt.savefig` to save an image in SVG format. This is because the `RenderSVG._make_id` method takes the first 10 characters of a `hashlib.md5` digest of entries in the SVG as the ID for entries in the SVG file. 

The Python docs for [hashlib](https://docs.python.org/3/library/hashlib.html) includes the following note:

> Constructors for hash algorithms that are always present in this module are sha1(), sha224(), sha256(), sha384(), sha512(), blake2b(), and blake2s(). md5() is normally available as well, though it may be missing if you are using a rare “FIPS compliant” build of Python.

## PR Checklist

- [na ] Has Pytest style unit tests
       - Not sure if this is applicable to my change
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [na] New features are documented, with examples if plot related 
- [x] Documentation is sphinx and numpydoc compliant
- [na] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
       - Not a major feature
- [x] Documented in doc/api/next_api_changes/* if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by daytonb** [comment]
Ok. I can update that and add it to the pull request.

**Comment by daytonb** [comment]
Alright I've added a file to the `doc/api/next_api_changes/behavior`. Let me know if the description seems unclear and needs tweaking.

File: doc/api/next_api_changes/behavior/18193-BKB.rst [hunk 1]
```diff
@@ -0,0 +1,11 @@
+ID attribute of XML tags in SVG files now based on SHA512 rather than MD5
```
Comment (QuLogic) [inline doc/api/next_api_changes/behavior/18193-BKB.rst:1]: But it's using `sha256`.

File: doc/api/next_api_changes/behavior/18193-BKB.rst [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The `RenderSVG._make_id` method generates unique ID attributes for various
```
Comment (QuLogic) [inline doc/api/next_api_changes/behavior/18193-BKB.rst:4]: I don't think we need to reference private functions. You should rewrite this to talk about SVG file saving only, and not internals.

File: doc/api/next_api_changes/behavior/18193-BKB.rst (hunk 1 above)
Comment (daytonb) [reply doc/api/next_api_changes/behavior/18193-BKB.rst:1]: Thanks! I had gone back and forth about which SHA to go for. I'm glad you caught this mistake!

File: doc/api/next_api_changes/behavior/18193-BKB.rst (hunk 2 above)
Comment (daytonb) [reply doc/api/next_api_changes/behavior/18193-BKB.rst:4]: OK. That sounds good to me.

**Comment by daytonb** [comment]
I re-wrote the behavior change note to no longer reference the private method. I was also previously getting an error from the circleci doc tests that I think was from the warnings Sphinx gave about my references to `hashlib.md5` and `RenderSVG._make_id`. The warnings said that Sphinx couldn't find these `pyobj`s to link to them. Obviously I removed the reference to `RenderSVG._make_id` so that won't be a problem any more, but I also reworded my note about `hashlib.md5` to avoid having to do a intersphinx reference.

**Comment by jkseppan** [comment]
Thanks @daytonb!

**Comment by daytonb** [comment]
It's been a pleasure!

**Pull request merged**